src/api/register_request.rs

```diff
@@ -0,0 +1,179 @@
+            let spec = spec.try_into()?;
+            api.pipeline.register_renderer(spec)?;
+            Ok(None)
+        }
+        RegisterRequest::WebRenderer(spec) => {
+            let spec = spec.try_into()?;
+            api.pipeline.register_renderer(spec)?;
+            Ok(None)
+        }
+        RegisterRequest::Image(spec) => {
+            let spec = spec.try_into()?;
+            api.pipeline.register_renderer(spec)?;
+            Ok(None)
+        }
+    }
+}
+
+fn register_output(api: &mut Api, request: RegisterOutputRequest) -> Result<(), ApiError> {
+    let RegisterOutputRequest {
+        output_id,
+        port,
+        resolution,
+        encoder_settings,
+        ip,
+    } = request;
+
+    api.pipeline.with_outputs(|mut iter| {
+        if let Some((node_id, _)) = iter.find(|(_, output)| output.port == port && output.ip == ip) {
+            return Err(ApiError::new(
+                "PORT_AND_IP_ALREADY_IN_USE",
+                format!("Failed to register output stream \"{output_id}\". Combination of port {port} and IP {ip} is already used by node \"{node_id}\""),
+                tiny_http::StatusCode(400)
+            ));
+        };
+        Ok(())
+    })?;
+
+    api.pipeline.register_output(
+        output_id.into(),
+        pipeline::OutputOptions {
+            resolution: resolution.into(),
+            encoder_settings: encoder_settings.into(),
+            receiver_options: rtp_sender::Options { port, ip },
+        },
+    )?;
+
+    Ok(())
+}
+
+fn register_input(
+    api: &mut Api,
+    request: RegisterInputRequest,
+) -> Result<ResponseHandler, ApiError> {
+    let RegisterInputRequest { input_id: id, port } = request;
+    let port: Port = port.try_into()?;
+
+    match port {
+        Port::Range((start, end)) => {
+            for port in start..=end {
+                trace!("[input {id}] checking port {port}");
+
+                if api
+                    .pipeline
+                    .inputs()
+                    .any(|(_, input)| input.port == port || input.port + 1 == port)
+                {
+                    trace!("[input {id}] port {port} is already used by another input",);
+                    continue;
+                }
+
+                let result = api.pipeline.register_input(
+                    id.clone().into(),
+                    rtp_receiver::Options {
+                        port,
+                        input_id: id.clone().into(),
+                    },
+                );
+
+                if check_port_not_available(&result, port).is_err() {
+                    trace!(
+                        "[input {id}] FFmpeg reported port registration failure for port {port}",
+                    );
+                    continue;
+                }
+
+                result?;
+
+                trace!("[input {id}] port registration succeeded for port {port}",);
+
+                return Ok(ResponseHandler::Response(Response::RegisteredPort(port)));
```

Comment:
```suggestion
                return match result {
                    Ok(_) => {
                        trace!("[input {id}] port registration succeeded for port {port}",);
                        Ok(ResponseHandler::Response(Response::RegisteredPort(port)));
                    }
                    Err(err) => Err(err);
                }
```
It's subjective, so it's up to you, but  I think sth like this is more readable. (the same for Exact variant)